tests/plugins/test_deepspeed_plugin.py

```diff
@@ -830,3 +830,55 @@ def test_deepspeed_multigpu_no_schedulers(tmpdir):
     trainer.fit(model)
 
     _assert_save_model_is_equal(model, tmpdir, trainer)
+
+
+@RunIf(min_gpus=1, deepspeed=True, special=True)
+def test_deepspeed_warn_train_dataloader_called(tmpdir):
+    """
+    Test DeepSpeed warns when it calls ``train_dataloader`` internally for logging batch size.
+    """
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        plugins=[DeepSpeedPlugin()],
+        gpus=1,
+        fast_dev_run=True,
+    )
+    with pytest.warns(UserWarning, match="Inferring the batch size for internal deepspeed logging"):
+        trainer.fit(model)
+
+
+@RunIf(min_gpus=1, deepspeed=True, special=True)
+def test_deepspeed_setup_train_dataloader(tmpdir):
+    """
+    Test DeepSpeed works when setup is required to call, and the user passes the batch size manually.
+    """
+
+    class PlDataModule(LightningDataModule):
+        def __init__(self):
+            super().__init__()
+            self._setup = False
+
+        def setup(self, stage: Optional[str] = None) -> None:
+            self._setup = True
+
+        def train_dataloader(self):
+            assert self._setup
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
```

Comment:
I think you want to test here that the dataloader() method gets called after setup when auto_select_batch_size gets used. 

But shouldn't that method better use the `trainer.train_dataloader`? Calling a LightningModule's dataloader method  several times may have unintended side effects.

Reply:
I think the issue arises from the logic being tested here living in the pre_training routine, before the dataloader is attached to the trainer (within the fit loop advance).

Reply:
ah I see, the call is here https://github.com/PyTorchLightning/pytorch-lightning/blob/b046bd0670cd86300ad77533d300db79b137054a/pytorch_lightning/trainer/trainer.py#L1127 but that's too late. 

shame :(